### PR TITLE
Particles: Remove particles on collision.

### PR DESCRIPTION
**Pull request by sofar**
Some particles are hard to use right now since they either go through
solid blocks (without collision detection), and with collision detection
enabled they (e.g. raindrops) would just stop dead on the floor and
sit there until they expire.

We can solve the problem by adding a boolean flag that tells the
particle to be removed if it ever collides with something. This will
make it easier to add rain that doesn't fall through your roof or stick
on the top of it. Or clouds and smoke that don't go through trees.

- [x] - ~~Verify that particle remove() is correct, I see way too many
remove events, so I don't think it's correct.~~ problem solved by force expiring particles.
- [x] - documentation - lua_api.txt

### CODE REVIEW

**Comment by ShadowNinja** [comment]
`collision_remove`.  Other flags weren't fixed because it was too late.

**Comment by sofar** [comment]
Alright, yes, renamed.

**Comment by 4aiman** [comment]
How about an experimental feature to call some method from Lua on collision? 
Maybe only for single particle (i.e. not for a particles spawner) to lessen possible performance impact and to somehow tell modder "Hey! Are you sure you want this?"..

Would be great to achieve some visual effects or making a weapon to shoot particles instead of creating objects.

**Comment by sofar** [comment]
@4aiman Particles are entirely client side, they can't create lua events for that reason.

**Comment by 4aiman** [comment]
@sofar I know that. That's why I suggested an experimental feature.
Client can calculate and send the coordinates where a particle will disappear to a server. Then the server may act.

Anyway, your comment left me thinking of possible use cases and now I clearly see this "is a no" due to security issues and/or cheating possibility.

**Comment by obneq** [comment]
@sofar hey instead of actively deleting the particle have you tried just setting its time beyond expiration so it gets deleted 'automatically' in the next step. might be simpler and work better, or maybe not, just an untested idea

**Comment by sofar** [comment]
@obneq thanks, that indeed worked well. I had to figure out how to do that, but now I'm seeing exactly 1 removal event per particle, and not dozens!

**Comment by sofar** [comment]
Both outstanding action items are now closed - this should be ready for review.

**Comment by neoascetic** [comment]
Want this!

**Comment by sofar** [comment]
@shadowninja @est31 @nerzhul review? This is working nicely now and it would be an appropriate time to merge.

File: doc/lua_api.txt [hunk 1]
```diff
@@ -3896,6 +3899,9 @@ Definition tables
     --  ^ minsize/maxsize, minexptime/maxexptime (expirationtime)
         collisiondetection = false,
     --  ^ collisiondetection: if true uses collision detection
+        collision_remove = false,
```
Comment (nerzhul) [inline doc/lua_api.txt]: i think collision_removal is better name as this is an attribute

File: src/particles.cpp [hunk 2]
```diff
@@ -131,12 +133,17 @@ void Particle::step(float dtime)
...
-		m_velocity = p_velocity/BS;
+		collisionMoveResult r = collisionMoveSimple(m_env,
+				m_gamedef, BS*0.5, box, 0,
+				dtime, &p_pos, &p_velocity,
+				m_acceleration * BS);
+		if (m_collision_remove && r.collides) {
+			// force expiration of the particle
+			m_expiration = -1.0;
+		} else {
+			m_pos = p_pos/BS;
```
Comment (nerzhul) [inline src/particles.cpp]: can you add spaces between the p_pos and BS for better reading ? same thing on the previous lines

File: src/server.cpp [hunk 3]
```diff
@@ -1695,7 +1697,8 @@ void Server::SendSpawnParticle(u16 peer_id, v3f pos, v3f velocity, v3f accelerat
 // Adds a ParticleSpawner on peer with peer_id
 void Server::SendAddParticleSpawner(u16 peer_id, u16 amount, float spawntime, v3f minpos, v3f maxpos,
 	v3f minvel, v3f maxvel, v3f minacc, v3f maxacc, float minexptime, float maxexptime,
-	float minsize, float maxsize, bool collisiondetection, bool vertical, std::string texture, u32 id)
+	float minsize, float maxsize, bool collisiondetection, bool collision_remove,
+	bool vertical, std::string texture, u32 id)
```
Comment (nerzhul) [inline src/server.cpp]: can you change this behavious by converting to const std::string& ?

File: src/server.h [hunk 4]
```diff
@@ -35,6 +35,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "chat_interface.h"
 #include "clientiface.h"
 #include "network/networkpacket.h"
+#include "collision.h"
```
Comment (nerzhul) [inline src/server.h]: this header is useless in server.h, please put it in cpp files

File: src/server.h [hunk 5]
```diff
@@ -275,15 +276,17 @@ class Server : public con::PeerHandler, public MapEventReceiver,
 	void spawnParticle(const std::string &playername,
 		v3f pos, v3f velocity, v3f acceleration,
```
Comment (nerzhul) [inline src/server.h]: also if you can fix those v3f to const v3f &

Comment (est31) [inline src/server.h]: I dont think that's required, copying is almost as expensive as a pointer.

File: src/particles.cpp [hunk 6]
```diff
@@ -131,12 +133,17 @@ void Particle::step(float dtime)
...
 		v3f p_pos = m_pos*BS;
 		v3f p_velocity = m_velocity*BS;
-		collisionMoveSimple(m_env, m_gamedef,
-			BS*0.5, box,
-			0, dtime,
-			&p_pos, &p_velocity, m_acceleration * BS);
-		m_pos = p_pos/BS;
-		m_velocity = p_velocity/BS;
+		collisionMoveResult r = collisionMoveSimple(m_env,
+				m_gamedef, BS*0.5, box, 0,
```
Comment (est31) [inline src/particles.cpp]: Indent with one tab only.

File: src/server.cpp (hunk 3 above)
Comment (est31) [inline src/server.cpp]: I agree with you @nerzhul  that passing std::string is bad because it does a wasteful, but can it be done in a separate PR, or at least a separate commit?

File: doc/lua_api.txt (hunk 1 above)
Comment (sofar) [inline doc/lua_api.txt]: Ok.

File: src/particles.cpp (hunk 2 above)
Comment (sofar) [inline src/particles.cpp]: Ok

File: src/particles.cpp (hunk 6 above)
Comment (sofar) [inline src/particles.cpp]: Ok.

File: src/server.h (hunk 4 above)
Comment (sofar) [inline src/server.h]: Yup, removed entirely as it wasn't used.

File: src/particles.cpp (hunk 6 above)
Comment (sofar) [inline src/particles.cpp]: ack, fixed.

**Comment by sofar** [comment]
Fixed up all comments, except for the `std::string` passing.

**Comment by sofar** [comment]
Pushed a separate commit to pass texture as `const std::string &`. Compiled and tested.

File: src/server.cpp (hunk 3 above)
Comment (sofar) [inline src/server.cpp]: Done in separate commit.

File: src/server.h (hunk 5 above)
Comment (kwolekr) [inline src/server.h]: unrelated changes shouldn't be made in the same PR

**Comment by neoascetic** [comment]
I want release 0.4.14 to contain these changes... :(
